src/pip/_internal/commands/install.py

```diff
@@ -594,6 +594,69 @@ def get_lib_location_guesses(*args, **kwargs):
     return [scheme['purelib'], scheme['platlib']]
 
 
+def site_packages_writable(**kwargs):
+    return all(
+        test_writable_dir(d) for d in set(get_lib_location_guesses(**kwargs))
+    )
+
+
+def decide_user_install(
+    use_user_site,  # type: Optional[bool]
+    prefix_path=None,  # type: Optional[str]
+    target_dir=None,  # type: Optional[str]
+    root_path=None,  # type: Optional[str]
+    isolated_mode=False,  # type: bool
+):
+    # type: (...) -> bool
+    """Determine whether to do a user install based on the input options.
+
+    If use_user_site is False, no additional checks are done.
+    If use_user_site is True, it is checked for compatibility with other
+    options.
+    If use_user_site is None, the default behaviour depends on the environment,
+    which is provided by the other arguments.
+    """
+    if use_user_site is False:
+        logger.debug("Non-user install by explicit request")
+        return False
+
+    if use_user_site is True:
+        if prefix_path:
```

Comment:
I know it's just transplanted from the existing logic, but I wonder why we aren't also throwing if `--target` is provided.

Reply:
@pradyunsg opened issue #7249 to consider this. Since it's the existing behaviour, I intend to leave it as is in this PR.